### ISSUE TITLE
Considering directly accessing GPU?

### ISSUE THREAD

**Issue by elliott-wen** [anbox/anbox#770]
1. Please check that no similar bug is already reported. Have a look on the list of open bugs at https://github.com/anbox/anbox/issues
2. Make sure you are running the latest version of Anbox before reporting an issue. Update snap to latest: `snap refresh --devmode --edge anbox`
3. Make sure you have debug logs enabled:
`sudo snap set anbox debug.enable=true`
4. Reproduce the error while debug logs enabled.
5. Run the anbox logs collection utility and attach the tar file.
`sudo /snap/bin/anbox.collect-bug-info `

6. ** Please paste the result of `anbox system-info` below:**
```
[please paste printout of `anbox system-info` here]
```

**Please describe your problem:**
Currently all GPU access is done via a proxy-like OpenGL driver provided by the Android Goldfish emulator. The driver in the Android side packs all GPU instructions as well as data and send them through certain tunnels to the host side. The host unpacks the instructions and carries out the real execution. 

It works well but may not be efficient considering the LXC container can directly access the host GPU.
I suggest we may be able to reimplement the opengl driver for Anbox such that it can make use of the direct GPU access. It may not be that hard as Mesa in Linux provides Opengl ES2 support and we probably just need to write a wrapper module for anbox.

**What were you expecting?:**

**Additional info:**

**Comment by morphis** [comment]
@elliott-wen I thought about this when I designed and implemented Anbox quite a lot but finally came to the decision that keeping the Android container itself independent of the host operating system is the better alternative. This way can use the same Android container across all kinds of devices. For example the same Android image (build for ARM of couse) runs on Ubuntu Touch.

Adding direct GPU access to the container will first off give faster performance but second off it introduces various problems like we would loose support for any users using a closed-source GPU driver. Also having support for both models adds unnecessary complexity. ChromeOS uses this model but its much easier for them as they control the entire stack from top to bottom. We don't and this makes it harder for us and remains easier if the we can run the Android system against an abstracted HAL implementation.

The way of how we want to improve performance is by adding a shared-memory transport between the container and the host Anbox process. Experiments I did show that this is enough to make things as fast as we need them to have. Alongside this we need various other improvements on the graphics stack we would have to do anyway regardless if we add direct GPU support to the container or not.
